draft-ietf-ppm-dap.md

```diff
@@ -1307,6 +1313,10 @@ of the task. The job ID is a 16-byte value, structured as follows:
 opaque AggregationJobID[16];
 ~~~
 
+The set of candidate reports MUST produce a set of report ID ranges which does
```

Comment:
I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Reply:
To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Reply:
> I agree that calling the new thing a "report ID range" makes everything hard to read. This sentence looks like it is comparing report ID ranges but it is actually comparing "report ID ranges".

Indeed; I am careful to use the term "interval" where an "interval" is truly needed. But "report ID range" does not capture that a "report ID range" includes a time window, too.

> Considering the fact that you're saying "a set of" here. Does that mean the old behavior would still be allowed by making each range have exactly one report ID?

Not quite -- an aggregation job may include reports in multiple time windows. In this case, there would be one "report ID range" for each time window represented in the aggregation job.

> To clarify, I am wondering why anybody would want to do this given that the report IDs are randomly assigned. Is the intent to allow people to recover from some amount of late arriving reports?

Nope, I think late-arriving reports are effectively lost (with high probability) in this scheme. The "report ID ranges" in use are all "maximal" in the sense that they stretch from the smallest to the largest available report ID.

Reply:
I think if the "set" only covers the "different time intervals" part you need to clarify. The way it's written the ranges don't need to be maximal. Somebody could send 3 ranges for t1 and 7 ranges for t2 as longs as they in total cover all the reports. And in the limit, send one range per report and everything turns into linear search again.

Reply:
I think the ranges do need to be maximal; there actually isn't much wiggle room here. 

The relevant Leader text is:

> The set of candidate reports MUST produce a set of report ID commitments which
does not conflict with the report ID commitments for any other aggregation job
associated with the task (see {{report-id-commitments}}).

[note that "report ID range" got renamed to "report ID commitment" here]

The relevant Helper text is:

> 1. Check if the report has been previously aggregated. If so, the input share
   MUST be marked as invalid with the error `report_replayed`. A report is
   considered previously aggregated if it conflicts with a report ID commitment
   for a completed aggregation job (see {{report-id-commitments}}).

That is, both fall back to the report-id-commitments section; this section specifies a maximal range:

> The report ID commitments associated with a given set of reports are determined by the following procedure:
> 
> 1. Partition the set of reports into groups based on the time window their
>    client timestamp falls into; this might be achieved by partitioning the
>    reports on a key generated by truncating each report's timestamp down to the
>    nearest multiple of the task's `time_precision`.
> 2. For each partitioned group of reports, generate an interval from the smallest
>    report ID in the group to the largest report ID in the group, inclusive. The
>    report ID commitment for this partition is this interval, associated with the
>    time window that generated this partition.

I had previously believed there was some wiggle room for what the Helper commits if the Leader fails in its `MUST` & sends an aggregation job containing a report which falls into an existing report ID commitment. However, now I think the Helper's behavior is fully specified here with no wiggle room:

* Any reports conflicting with an existing report ID commitment must be marked as `report_replayed`.
* The new commitment associated with the aggregation job is still the maximal report ID commitments suggested by the report-id-commitments section -- this commitment will conflict with an already-existing commitment, but that's OK. (A clever implementation might merge conflicting commitments without changing its externally-observable behavior, but this isn't required.)

Reply:
By the way -- I'm interested in your thoughts on the new text in the report-id-commitments section @simon-friedberger. Other than renaming "report ID ranges" to "report ID commitments", I tried to make it clearer that the set of report ID commitments is one-per-time-window, rather than allowing the aggregators to generate a larger count of smaller commitment intervals.